lib/rouge/lexers/php.rb

```diff
@@ -137,12 +143,17 @@ def self.detect?(text)
         end
 
         rule %r/(true|false|null)\b/, Keyword::Constant
+        rule /(?:void|\??(?:int|float|bool|string|iterable))\b/i, Keyword::Type
+        rule /\??(?:self|callable)\b/i, Keyword::Type
         rule %r/\$\{\$+#{id}\}/i, Name::Variable
         rule %r/\$+#{id}/i, Name::Variable
+        rule /(yield)(#{WHITESPACE})(from)/io do
+          groups Keyword, Text, Keyword
+        end
 
         # may be intercepted for builtin highlighting
         rule %r/\\?#{nsid}/i do |m|
-          name = m[0]
+          name = m[0].downcase
```

Comment:
If you were to create a new memoised function called `self.types` and save the types you'd put in earlier rules there, would you be able to eliminate the rules above and check for matches down here. For memory reasons, it's preferable to do that.

Reply:
If I got your point: no. A real handling of typehinting would be great but you can't do it from a lexer. So I just recognize the few builtin types.

Reply:
Sorry, what I meant specifically was whether you could create a memoising function called `types` that listed the built-in types you've put into this regex: `%r/(?:void|\??(?:int|float|bool|string|iterable|self|callable))\b/i`.

Reply:
PHP has this very few builtin types (only these) so isn't a specific rule a better fit?

Also, without a rule, you'd need to do an extra lookup in `rule %r/\\?#{nsid}/i` and it becomes trickier to handle the front `?` (which means that `NULL` is allowed) with the exception of `void`, right?

Reply:
Yeah, I see what you mean. Got it.